### PR TITLE
ES example

### PR DESCRIPTION
**Pull request by zenghsh3**
- add ES example
- fix parl.compile bug
- avoid `set_weights` function call `is_gpu_available` repeatedly

About 100 minutes can reach 6000 scores in Humanoid. (96 CPUs)
image

### CODE REVIEW

File: examples/ES/mujoco_model.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class MujocoModel(parl.Model):
+    def __init__(self, act_dim):
+        hid1_size = 256
+        hid2_size = 256
+
+        self.fc1 = layers.fc(size=hid1_size, act='tanh')
+        self.fc2 = layers.fc(size=hid2_size, act='tanh')
+        self.fc3 = layers.fc(size=act_dim, act=None)
```
Comment (TomorrowIsAnOtherDay) [inline examples/ES/mujoco_model.py]: remove `act=None`

File: examples/ES/noise.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Third party code
+#
+# The following code are copied or modified from:
+# https://github.com/ray-project/ray/blob/master/python/ray/rllib/utils/filter.py
+
+import numpy as np
+
+
+class SharedNoiseTable(object):
+    """
```
Comment (TomorrowIsAnOtherDay) [inline examples/ES/noise.py]: Pls add more comment to describe its definition and usage.

File: examples/ES/noise.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        return noise
+
+    def get(self, i, dim):
+        return self.noise[i:i + dim]
+
+    def sample_index(self, dim):
+        return np.random.randint(0, len(self.noise) - dim + 1)
+
+
+if __name__ == '__main__':
```
Comment (TomorrowIsAnOtherDay) [inline examples/ES/noise.py]: Are the following lines used for test ? Please delete them if True.

File: examples/ES/obs_filter.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    def std(self):
+        return np.sqrt(self.var)
+
+    @property
+    def shape(self):
+        return self._M.shape
+
+
+class MeanStdFilter(Filter):
+    """Keeps track of a running mean for seen states"""
```
Comment (TomorrowIsAnOtherDay) [inline examples/ES/obs_filter.py]: Please give mode detail about its usage.

**Review by TomorrowIsAnOtherDay — CHANGES_REQUESTED** [review]

File: examples/ES/learner.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+        # Send set_filter signal to all actors
+        self.latest_obs_filter = self.obs_filter.as_serializable()
+        for q in self.actors_signal_input_queues:
+            q.put({'signal': 'set_filter'})
+
+    def log_metrics(self, metrics):
+        logger.info(metrics)
+        self.csv_logger.log_dict(metrics)
+        for k, v in metrics.items():
```
Comment (TomorrowIsAnOtherDay) [inline examples/ES/learner.py:206]: Before adding a new data point, pls check if it is None.

File: examples/ES/mujoco_model.py (hunk 1 above)
Comment (zenghsh3) [reply examples/ES/mujoco_model.py]: done

File: examples/ES/noise.py (hunk 2 above)
Comment (zenghsh3) [reply examples/ES/noise.py]: done

File: examples/ES/noise.py (hunk 3 above)
Comment (zenghsh3) [reply examples/ES/noise.py]: done

File: examples/ES/obs_filter.py (hunk 4 above)
Comment (zenghsh3) [reply examples/ES/obs_filter.py]: done

File: examples/ES/learner.py (hunk 5 above)
Comment (zenghsh3) [reply examples/ES/learner.py:206]: done

File: examples/ES/learner.py [hunk 6]
```diff
@@ -206,9 +203,6 @@ def _update_filter(self):
 
     def log_metrics(self, metrics):
         logger.info(metrics)
-        self.csv_logger.log_dict(metrics)
         for k, v in metrics.items():
-            tensorboard.add_scalar(k, v, self.sample_total_steps)
-
-    def close(self):
-        self.csv_logger.close()
+            if value is not None:
```
Comment (TomorrowIsAnOtherDay) [inline examples/ES/learner.py]: where is the definition of `value`? It should be `v`

Comment (zenghsh3) [reply examples/ES/learner.py]: fixed

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
Looks good to me.

**Pull request merged**